### PR TITLE
linode-marketplace-jupyterlab

### PR DESCRIPTION
**Pull request by ericruzanski**
Adding JupyterLab to the Akamai Connected Cloud Marketplace.

### CODE REVIEW

File: apps/linode-marketplace-jupyterlab/provision.yml
Comment (tbaka) [inline apps/linode-marketplace-jupyterlab/provision.yml:1]: For new releases we will be generating sudo user credentials in provision.yml. You can add
``` 
tasks:
    - name: generating sudo user and password 
      import_tasks: ../linode_helpers/roles/sudouser/tasks/main.yml
```

File: apps/linode-marketplace-jupyterlab/roles/common/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+# dns creation
+- name: create dns A record
+  import_role:
+    name: create_dns_record
+  when: 
+    - token_password is defined
+    - default_dns is not defined
+
+# set customer supplied sudo user, password and SSH keys
+- name: set sudo user
```
Comment (tbaka) [inline apps/linode-marketplace-jupyterlab/roles/common/tasks/main.yml:25]: remove from common. add to provision

File: apps/linode-marketplace-jupyterlab/roles/jupyterlab/templates/motd.j2 [hunk 2]
```diff
@@ -0,0 +1,7 @@
+*********************************************************
+Akamai Connected Cloud JupyterLab Marketplace App
+App URL: https://{{ _domain }}
+Obtaining the JupyterLab Token: ./get_jupyter_token.sh
```
Comment (tbaka) [inline apps/linode-marketplace-jupyterlab/roles/jupyterlab/templates/motd.j2:4]: creds are being standardized into `/root/.credentials`

File: apps/linode-marketplace-jupyterlab/site.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  connection: local
+  gather_facts: true
+  any_errors_fatal: true
+  user: root
+  vars_files:
+    - group_vars/linode/vars
+  roles:
+    - common
+    - docker
+    - jupyterlab
```
Comment (tbaka) [inline apps/linode-marketplace-jupyterlab/site.yml:14]: add `- post` task to write creds. 
example `post/tasks/main.yml`
```
# motd and creds gen
- name: copy MOTD template to /etc/motd
  template:
    src: templates/motd.j2
    dest: /etc/motd

- name: writing docker creds into file
  copy:
    dest: /root/.credentials
    mode: '0600'
    content: |
      Sudo Username: {{ username }}
      Sudo Password: {{ password }}
```

File: deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh [hunk 4]
```diff
@@ -0,0 +1,119 @@
+#!/bin/bash
+set -e
+trap "cleanup $? $LINENO" EXIT
+
+## JupyterLab Settings
+#<UDF name="soa_email_address" label="Email address (for the Let's Encrypt SSL certificate)" example="user@domain.tld">
+
+## Linode/SSH Security Settings
+#<UDF name="username" label="The limited sudo user to be created for the Linode" default="">
```
Comment (tbaka) [inline deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh:9]: make this a required field by removing the `default=""`

File: deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    rm -rf ${WORK_DIR}
+  fi
+
+}
+
+function udf {
+  local group_vars="${WORK_DIR}/${MARKETPLACE_APP}/group_vars/linode/vars"
+  echo "webserver_stack: lemp" >> ${group_vars};
+  
+  if [[ -n ${USER_NAME} ]]; then
```
Comment (tbaka) [inline deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh:38]: can be removed.

File: deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  fi
+ 
+  if [[ -n ${TOKEN_PASSWORD} ]]; then
+    echo "token_password: ${TOKEN_PASSWORD}" >> ${group_vars};
+  else echo "No API token entered";
+  fi
+
+}
+
+function create_jupyter_token_script {
```
Comment (tbaka) [inline deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh:82]: capturing creds should be moved into `/roles/post/tasks/main.yml` and completed with Ansible.

**Review by tbaka — CHANGES_REQUESTED** [review]
This looks good and does pass deploy. I'm going to request some changes because we are updating a few standards. Sudo users will be required, and passwords will be generated then provided to the end-user, and all credentials will be captured in a standard `post` play. 

Let me know if this is clear or if you have any questions.

File: apps/linode-marketplace-jupyterlab/provision.yml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    - group_vars/linode/vars
+  tasks:
+    - name: generating sudo user and password
+      import_tasks: ../linode_helpers/roles/sudouser/tasks/main.yml
+
+# set sudo user, password
+- name: set sudo user
+  import_role:
+    name: sudouser
+  when: username is defined
```
Comment (tbaka) [inline apps/linode-marketplace-jupyterlab/provision.yml:17]:
```suggestion

```
import_task above will fire 100% of the time.

File: deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+#<UDF name="pubkey" label="The SSH Public Key that will be used to access the Linode" default="">
+#<UDF name="disable_root" label="Disable root access over SSH?" oneOf="Yes,No" default="No">
+
+## Domain Settings
+#<UDF name="token_password" label="Your Linode API token. This is needed to create your server's DNS records" default="">
+#<UDF name="subdomain" label="Subdomain" example="The subdomain for the DNS record: www (Requires Domain)" default="">
+#<UDF name="domain" label="Domain" example="The domain for the DNS record: example.com (Requires API token)" default="">
+
+# git repo
+export GIT_REPO="https://github.com/ericruzanski/marketplace-apps.git"
```
Comment (tbaka) [inline deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh:19]:
```suggestion
export GIT_REPO="https://github.com/akamai-compute-marketplace/marketplace-apps.git"
```

File: deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+function udf {
+  local group_vars="${WORK_DIR}/${MARKETPLACE_APP}/group_vars/linode/vars"
+  echo "webserver_stack: lemp" >> ${group_vars};
+  
+  if [[ -n ${USER_NAME} ]]; then
+    echo "username: ${USER_NAME}" >> ${group_vars};
+  else echo "No username entered";
+  fi
```
Comment (tbaka) [inline deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh]:
```suggestion

```

File: deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  if [ -d "${WORK_DIR}" ]; then
+    rm -rf ${WORK_DIR}
+  fi
+
+}
+
+function udf {
+  local group_vars="${WORK_DIR}/${MARKETPLACE_APP}/group_vars/linode/vars"
+  echo "webserver_stack: lemp" >> ${group_vars};
+
```
Comment (tbaka) [inline deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh:36]:
```suggestion
  sed 's/  //g' <<EOF > ${group_vars}

  # sudo username
  username: ${USER_NAME}
  webserver_stack: lemp
  EOF
```

Comment (tbaka) [reply deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh:36]: we hadn't decided on this sed convention when I requested changes.

File: apps/linode-marketplace-jupyterlab/roles/post/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+      Sudo Username: {{ username }}
+      Sudo Password: {{ password }}
+
+# Retrieving and storing Jupyter token
+- name: Retrieve and store Jupyter token
+  command: "{{ item }}"
+  loop:
+    - "docker logs jupyter_jupyterlab_1 2>&1 | grep -oP '(?<=token=)[^ ]+' | head -n 1 >> /root/.credentials"
+    - "chmod 0600 /root/.credentials"
+  ignore_errors: true
```
Comment (tbaka) [inline apps/linode-marketplace-jupyterlab/roles/post/tasks/main.yml:21]: It is generally best practice to always use Ansible modules / builtins if you can vs. running a straight shell command through it. 
I would recommend rewriting this command loop into a couple of tasks so you can use Ansible. Something like: 
```
- name: std out docker logs consul 
  ansible.builtin.shell: docker logs jupyter_jupyterlab_1
  register: result
 # until handles docker catching up with ansible. something like this is preferable to ignore_errors true
  until: "result is not failed"
  retries: 3
  delay: 3

- name: set jupyter token fact 
  ansible.builtin.set_fact: 
    jupyter_token: "{{ result.stdout | regex_search('(?<=token=)[^ ]+') }}"
    cacheable: yes

- name: writing creds into file
  copy:
    dest: /root/.credentials
    mode: '0600'
    content: |
      Sudo Username: {{ username }}
      Sudo Password: {{ password }}
      Jupyter Token: {{ jupyter_token }}
```

File: deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh (hunk 8 above)
Comment (ericruzanski) [reply deployment_scripts/linode-marketplace-jupyterlab/jupyterlab-deploy.sh:19]: Will flip this back once PR changes pass test deployment.

**Comment by tbaka** [comment]
closed in favor of new PR

**Pull request closed without merging**